### PR TITLE
Refactor X86InductorQuantizer Conv2D annotation with SubgraphMatcherWithNameNodeMap

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #116688
* __->__ #116413

**Summary**
As per the discussion with Jerry in https://github.com/pytorch/pytorch/pull/114540#discussion_r1406945902, we plan to refactor `X86InductorQuantizer` annotations with `SubgraphMatcherWithNameNodeMap`. This plan will be implemented through several PRs, with this one being the first, focusing on refactoring single conv2d annotations in PTQ.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @aakhundov

### CODE REVIEW

**Comment by jerryzh168** [comment]
great! maybe you can test different conv configurations to make sure the pattern can catch all of them, including different kernel_size/stride/padding and symbolic shapes

File: torch/ao/quantization/quantizer/utils.py [hunk 1]
```diff
@@ -47,3 +60,55 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
...
+    return SubgraphMatcherWithNameNodeMap(pattern, ignore_literals=True)
+
+
+@functools.lru_cache(None)
+def _generate_conv2d_pattern_matcher():
+    # Ensure it's only be invoked once, due to the cache size limitation in
+    # https://github.com/pytorch/pytorch/blob/
+    # 4c6e842496da636123f83ef868ca1974631f1f1e/torch/_dynamo/config.py#L35-L39
+    return _generate_pattern_matcher_helper(
+        _get_conv_unary_pattern(torch.nn.functional.conv2d),
```
Comment (jgong5) [inline torch/ao/quantization/quantizer/utils.py:112]: How is `torch.nn.Conv2d` handled?

Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/utils.py:112]: I think `torch.nn.Conv2d` and `torch.nn.functional.conv2d` will generate same node `torch.ops.aten.conv2d.default` after export. So, we should be fine to only use `torch.nn.functional.conv2d` here.

**Comment by leslie-fang-intel** [comment]
> great! maybe you can test different conv configurations to make sure the pattern can catch all of them, including different kernel_size/stride/padding and symbolic shapes

- Add the testing for different `kernel_size/stride/padding`
- For `symbolic shapes`, did you mean add `constraints` parameters in `capture_pre_autograd_graph`? I think there should be no `symbolic size` for the pattern generated by Conv2D.

**Review by jgong5 — APPROVED** [review]

**Comment by jerryzh168** [comment]
> * For `symbolic shapes`, did you mean add `constraints` parameters in `capture_pre_autograd_graph`? I think there should be no `symbolic size` for the pattern generated by Conv2D.

Yeah, but we may still want to test it to make sure it also works

File: torch/ao/quantization/quantizer/utils.py [hunk 2]
```diff
@@ -47,3 +60,55 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
         ((user not in partition_nodes) or _is_sym_size_node(user))
         for user in node.users
     )
+
+
+def _get_conv_unary_pattern(
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/utils.py:65]: can you add some comments for this

**Review by jerryzh168 — APPROVED** [review]

File: torch/ao/quantization/quantizer/utils.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/utils.py:65]: Thanks and added.

File: torch/ao/quantization/quantizer/utils.py [hunk 3]
```diff
@@ -47,3 +60,57 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
...
+    conv_fn: Callable,
+    has_bn: bool = False,  # Usually need for QAT pattern
+    unary_fn: Optional[Callable[[Any], Any]] = None,
+):
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/utils.py:69]: can you add a return type as well

File: torch/ao/quantization/quantizer/utils.py [hunk 4]
```diff
@@ -47,3 +60,57 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
...
+            "conv": conv,
+            "conv_weight": conv_weight,
+            "conv_bias": conv_bias,
+            "output": output,
+        }
+
+    return _conv_unary
+
+
+def _generate_pattern_matcher_helper(pattern, example_inputs):
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/utils.py:101]: nit: type annotations

File: torch/ao/quantization/quantizer/utils.py [hunk 5]
```diff
@@ -47,3 +60,57 @@ def _node_only_used_for_sym_size(node: Node, partition_nodes: List[Node]):
...
+    pattern = get_aten_graph_module(pattern, example_inputs)
+    pattern.graph.eliminate_dead_code()
+    pattern.recompile()
+    return SubgraphMatcherWithNameNodeMap(pattern, ignore_literals=True)
+
+
+@functools.lru_cache(None)
+def _generate_conv2d_pattern_matcher():
```
Comment (jerryzh168) [inline torch/ao/quantization/quantizer/utils.py:109]: nit: type annotations

File: torch/ao/quantization/quantizer/utils.py (hunk 3 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/utils.py:69]: Thanks for the comment and added.

File: torch/ao/quantization/quantizer/utils.py (hunk 4 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/utils.py:101]: Thanks and added.

File: torch/ao/quantization/quantizer/utils.py (hunk 5 above)
Comment (leslie-fang-intel) [reply torch/ao/quantization/quantizer/utils.py]: Thanks and added.
